core/src/main/java/org/jdbi/v3/core/cache/internal/InternalJdbiCache.java

```diff
@@ -0,0 +1,167 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core.cache.internal;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ConcurrentMap;
+
+import org.jdbi.v3.core.cache.JdbiCache;
+import org.jdbi.v3.core.cache.JdbiCacheBuilder;
+import org.jdbi.v3.core.cache.JdbiCacheLoader;
+
+public final class InternalJdbiCache<K, V> implements JdbiCache<K, V> {
+
+    private final ConcurrentMap<K, V> cache;
+    private final ConcurrentLinkedQueue<K> expungeQueue;
+    private final JdbiCacheLoader<K, V> cacheLoader;
+
+    private final int maxSize;
+
+    private InternalJdbiCache(Builder builder, JdbiCacheLoader<K, V> cacheLoader) {
+        this.cache = new ConcurrentHashMap<>();
+        this.expungeQueue = new ConcurrentLinkedQueue<>();
+        this.cacheLoader = cacheLoader;
+
+        this.maxSize = builder.maxSize;
+    }
+
+    @Override
+    public V get(K key) {
+        try {
+            if (cacheLoader == null) {
+                return cache.get(key);
+            } else {
+                return cache.computeIfAbsent(key, compute(cacheLoader)::create);
+            }
+        } finally {
+            expunge();
+        }
+    }
+
+    @Override
+    public V getWithLoader(K key, JdbiCacheLoader<K, V> cacheLoader) {
+        try {
+            return cache.computeIfAbsent(key, compute(cacheLoader)::create);
+        } finally {
+            expunge();
+        }
+    }
+
+    @Override
+    public CacheStats getStats() {
+        return new CacheStats(cache.size(), maxSize);
+    }
+
+    private JdbiCacheLoader<K, V> compute(JdbiCacheLoader<K, V> delegate) {
+        if (delegate == null) {
+            return null;
+        }
+        if (maxSize <= 0) {
+            return delegate;
+        }
+
+        return key -> {
+            V value = delegate.create(key);
+            if (value != null) {
+                expungeQueue.add(key);
+            }
+            return value;
+        };
+    }
+
+    private void expunge() {
+        if (maxSize <= 0) {
+            // unbounded
+            return;
+        }
+
+        // quick and dirty implementation that runs in
+        // linear time to remove a few elements from the cache
+        // if the cache really exceeded its size. This is a simple
+        // FIFO. Anything more sophisticated should probably use a better
+        // implementation of the cache code.
+        int cacheSize = expungeQueue.size();
```

Comment:
fyi this is an O(n) call. You might use the map which maintains it in a counter (LongAddder).

> Beware that, unlike in most collections, this method is NOT a constant-time operation.

Reply:
Hey ben. Thank you for the comment. Yes, it is and it is called only once. Alternative would be to maintain a counter on the side. I will look into adding this.

Reply:
Called once per read currently, which could be surprisingly expensive.

Reply:
I added an int to track the cache size, increase it when the cache loader returns a non-null value and use that to determine the size of the cache and expunge it. So now everything should be O(1).